File: office_plugin/hosts/OleFormulaObjectNative/src/FormulaOleObject.cpp
#include "FormulaOleObject.h"

#include "NativeLog.h"
#include "OleFormulaIds.h"
#include "PendingPayload.h"
#include "StorageUtil.h"
#include "Win32Check.h"

#include <atlconv.h>
#include <comdef.h>
#include <new>
#include <shellapi.h>

namespace
{
volatile LONG g_objectCount = 0;
volatile LONG g_lockCount = 0;

HRESULT ValidateContentAspect(DWORD aspect)
{
    return aspect == DVASPECT_CONTENT ? S_OK : DV_E_DVASPECT;
}
}

LONG GetNativeOleObjectCount()
{
    return g_objectCount;
}

LONG GetNativeOleLockCount()
{
    return g_lockCount;
}

FormulaOleObject::FormulaOleObject()
    : presentation_(CreatePresentationFromPayload(ConsumePendingPayload()))
{
    if (presentation_.latex.empty())
    {
        presentation_ = CreatePlaceholderPresentation(kFormulaDefaultLatex);
    }

    WriteNativeOleLog(L"FormulaOleObject constructed.");
    InterlockedIncrement(&g_objectCount);
}

FormulaOleObject::~FormulaOleObject()
{
    WriteNativeOleLog(L"FormulaOleObject destructed.");
    InterlockedDecrement(&g_objectCount);
}

STDMETHODIMP FormulaOleObject::QueryInterface(REFIID iid, void** object)
{
    if (object == nullptr)
    {
        return E_POINTER;
    }

    if (iid == IID_IUnknown || iid == IID_IOleObject)
    {
        *object = static_cast<IOleObject*>(this);
    }
    else if (iid == IID_IDataObject)
    {
        *object = static_cast<IDataObject*>(this);
    }
    else if (iid == IID_IViewObject)
    {
        *object = static_cast<IViewObject*>(this);
    }
    else if (iid == IID_IPersist || iid == IID_IPersistStorage)
    {
        *object = static_cast<IPersistStorage*>(this);
    }
    else
    {
        *object = nullptr;
        return E_NOINTERFACE;
    }

    AddRef();
    return S_OK;
}

STDMETHODIMP_(ULONG) FormulaOleObject::AddRef()
{
    return static_cast<ULONG>(InterlockedIncrement(&refCount_));
}

STDMETHODIMP_(ULONG) FormulaOleObject::Release()
{
    const ULONG remaining = static_cast<ULONG>(InterlockedDecrement(&refCount_));
    if (remaining == 0)
    {
        delete this;
    }

    return remaining;
}

STDMETHODIMP FormulaOleObject::SetClientSite(IOleClientSite* clientSite)
{
    clientSite_ = clientSite;
    return S_OK;
}

STDMETHODIMP FormulaOleObject::GetClientSite(IOleClientSite** clientSite)
{
    if (clientSite == nullptr)
    {
        return E_POINTER;
    }

    return clientSite_.CopyTo(clientSite);
}

STDMETHODIMP FormulaOleObject::SetHostNames(LPCOLESTR, LPCOLESTR)
{
    return S_OK;
}

STDMETHODIMP FormulaOleObject::Close(DWORD)
{
    clientSite_.Release();
    return S_OK;
}

STDMETHODIMP FormulaOleObject::SetMoniker(DWORD, IMoniker*)
{
    return S_OK;
}

STDMETHODIMP FormulaOleObject::GetMoniker(DWORD, DWORD, IMoniker** moniker)
{
    if (moniker == nullptr)
    {
        return E_POINTER;
    }

    *moniker = nullptr;
    return E_NOTIMPL;
}

STDMETHODIMP FormulaOleObject::InitFromData(IDataObject*, BOOL, DWORD)
{
    return S_OK;
}

STDMETHODIMP FormulaOleObject::GetClipboardData(DWORD, IDataObject** dataObject)
{
    if (dataObject == nullptr)
    {
        return E_POINTER;
    }

    return QueryInterface(IID_IDataObject, reinterpret_cast<void**>(dataObject));
}

STDMETHODIMP FormulaOleObject::DoVerb(LONG verb, LPMSG, IOleClientSite*, LONG, HWND parentWindow, LPCRECT)
{
    if (verb == OLEIVERB_PRIMARY || verb == OLEIVERB_SHOW || verb == OLEIVERB_OPEN)
    {
        StoreEditorPayload(presentation_.payloadJson);
        std::wstring renderer = ResolveRendererPath();
        if (renderer.empty())
        {
            return OLEOBJ_S_CANNOT_DOVERB_NOW;
        }

        SHELLEXECUTEINFOW executeInfo{};
        executeInfo.cbSize = sizeof(executeInfo);
        executeInfo.fMask = SEE_MASK_NOCLOSEPROCESS;
        executeInfo.hwnd = parentWindow;
        executeInfo.lpVerb = L"open";
        executeInfo.lpFile = renderer.c_str();
        executeInfo.lpParameters = L"/EditPayload";
        executeInfo.nShow = SW_SHOWNORMAL;
        if (!ShellExecuteExW(&executeInfo))
        {
            return OLEOBJ_S_CANNOT_DOVERB_NOW;
        }

        if (executeInfo.hProcess != nullptr)
        {
            WaitForSingleObject(executeInfo.hProcess, INFINITE);
            CloseHandle(executeInfo.hProcess);
        }

        std::wstring updatedPayload = ConsumeEditorPayloadResult();
        if (updatedPayload.empty() || updatedPayload == presentation_.payloadJson)
        {
            return S_OK;
        }

        FormulaPresentation updatedPresentation = CreatePresentationFromPayload(updatedPayload);
        if (updatedPresentation.latex.empty())
        {
            return OLEOBJ_S_CANNOT_DOVERB_NOW;
        }

        presentation_ = std::move(updatedPresentation);
        dirty_ = true;
        if (viewAdviseSink_ != nullptr)
        {
            viewAdviseSink_->OnViewChange(DVASPECT_CONTENT, -1);
        }

        if (clientSite_ != nullptr)
        {
            clientSite_->SaveObject();
        }

        return S_OK;
    }

    return S_OK;
}

STDMETHODIMP FormulaOleObject::EnumVerbs(IEnumOLEVERB** enumOleVerb)
{
    if (enumOleVerb == nullptr)
    {
        return E_POINTER;
    }

    return OleRegEnumVerbs(CLSID_LaTeXSnipperFormula, enumOleVerb);
}

STDMETHODIMP FormulaOleObject::Update()
{
    return S_OK;
}

STDMETHODIMP FormulaOleObject::IsUpToDate()
{
    return S_OK;
}

STDMETHODIMP FormulaOleObject::GetUserClassID(CLSID* classId)
{
    if (classId == nullptr)
    {
        return E_POINTER;
    }

    *classId = CLSID_LaTeXSnipperFormula;
    return S_OK;
}

STDMETHODIMP FormulaOleObject::GetUserType(DWORD, LPOLESTR* userType)
{
    if (userType == nullptr)
    {
        return E_POINTER;
    }

    const size_t length = wcslen(kFormulaFriendlyName) + 1;
    *userType = static_cast<LPOLESTR>(CoTaskMemAlloc(length * sizeof(wchar_t)));
    if (*userType == nullptr)
    {
        return E_OUTOFMEMORY;
    }

    wcscpy_s(*userType, length, kFormulaFriendlyName);
    return S_OK;
}

STDMETHODIMP FormulaOleObject::SetExtent(DWORD drawAspect, SIZEL* size)
{
    if (size == nullptr)
    {
        return E_POINTER;
    }

    HRESULT aspectResult = ValidateContentAspect(drawAspect);
    if (FAILED(aspectResult))
    {
        return aspectResult;
    }

    presentation_.himetricSize = {size->cx, size->cy};
    return S_OK;
}

STDMETHODIMP FormulaOleObject::GetExtent(DWORD drawAspect, SIZEL* size)
{
    if (size == nullptr)
    {
        return E_POINTER;
    }

    HRESULT aspectResult = ValidateContentAspect(drawAspect);
    if (FAILED(aspectResult))
    {
        return aspectResult;
    }

    size->cx = presentation_.himetricSize.cx;
    size->cy = presentation_.himetricSize.cy;
    return S_OK;
}

STDMETHODIMP FormulaOleObject::Advise(IAdviseSink*, DWORD* connection)
{
    if (connection == nullptr)
    {
        return E_POINTER;
    }

    *connection = 0;
    return OLE_E_ADVISENOTSUPPORTED;
}

STDMETHODIMP FormulaOleObject::Unadvise(DWORD)
{
    return OLE_E_ADVISENOTSUPPORTED;
}

STDMETHODIMP FormulaOleObject::EnumAdvise(IEnumSTATDATA** enumAdvise)
{
    if (enumAdvise == nullptr)
    {
        return E_POINTER;
    }

    *enumAdvise = nullptr;
    return OLE_E_ADVISENOTSUPPORTED;
}

STDMETHODIMP FormulaOleObject::GetMiscStatus(DWORD aspect, DWORD* status)
{
    if (status == nullptr)
    {
        return E_POINTER;
    }

    HRESULT aspectResult = ValidateContentAspect(aspect);
    if (FAILED(aspectResult))
    {
        return aspectResult;
    }

    *status = OLEMISC_RECOMPOSEONRESIZE | OLEMISC_CANTLINKINSIDE | OLEMISC_INSIDEOUT;
    return S_OK;
}

STDMETHODIMP FormulaOleObject::SetColorScheme(LOGPALETTE*)
{
    return S_OK;
}

STDMETHODIMP FormulaOleObject::GetData(FORMATETC* format, STGMEDIUM* medium)
{
    if (format == nullptr || medium == nullptr)
    {
        return E_POINTER;
    }

    HRESULT queryResult = QueryGetData(format);
    if (FAILED(queryResult))
    {
        return queryResult;
    }

    HENHMETAFILE metafile = CopyEnhMetaFileFromBytes(presentation_.enhancedMetafile);
    if (metafile == nullptr)
    {
        return E_FAIL;
    }

    medium->tymed = TYMED_ENHMF;
    medium->hEnhMetaFile = metafile;
    medium->pUnkForRelease = nullptr;
    return S_OK;
}

STDMETHODIMP FormulaOleObject::GetDataHere(FORMATETC*, STGMEDIUM*)
{
    return DATA_E_FORMATETC;
}

STDMETHODIMP FormulaOleObject::QueryGetData(FORMATETC* format)
{
    if (format == nullptr)
    {
        return E_POINTER;
    }

    if (format->cfFormat != CF_ENHMETAFILE)
    {
        return DV_E_FORMATETC;
    }

    if ((format->tymed & TYMED_ENHMF) == 0)
    {
        return DV_E_TYMED;
    }

    return ValidateContentAspect(format->dwAspect);
}

STDMETHODIMP FormulaOleObject::GetCanonicalFormatEtc(FORMATETC*, FORMATETC* output)
{
    if (output == nullptr)
    {
        return E_POINTER;
    }

    ZeroMemory(output, sizeof(*output));
    output->ptd = nullptr;
    return DATA_S_SAMEFORMATETC;
}

STDMETHODIMP FormulaOleObject::SetData(FORMATETC*, STGMEDIUM*, BOOL)
{
    return E_NOTIMPL;
}

STDMETHODIMP FormulaOleObject::EnumFormatEtc(DWORD, IEnumFORMATETC** enumFormatEtc)
{
    if (enumFormatEtc == nullptr)
    {
        return E_POINTER;
    }

    *enumFormatEtc = nullptr;
    return E_NOTIMPL;
}

STDMETHODIMP FormulaOleObject::DAdvise(FORMATETC*, DWORD, IAdviseSink*, DWORD* connection)
{
    if (connection == nullptr)
    {
        return E_POINTER;
    }

    *connection = 0;
    return OLE_E_ADVISENOTSUPPORTED;
}

STDMETHODIMP FormulaOleObject::DUnadvise(DWORD)
{
    return OLE_E_ADVISENOTSUPPORTED;
}

STDMETHODIMP FormulaOleObject::EnumDAdvise(IEnumSTATDATA** enumAdvise)
{
    if (enumAdvise == nullptr)
    {
        return E_POINTER;
    }

    *enumAdvise = nullptr;
    return OLE_E_ADVISENOTSUPPORTED;
}

STDMETHODIMP FormulaOleObject::Draw(DWORD drawAspect, LONG, void*, DVTARGETDEVICE*, HDC, HDC drawContext, LPCRECTL bounds, LPCRECTL, BOOL(__stdcall*)(ULONG_PTR), ULONG_PTR)
{
    HRESULT aspectResult = ValidateContentAspect(drawAspect);
    if (FAILED(aspectResult))
    {
        return aspectResult;
    }

    if (drawContext == nullptr || bounds == nullptr)
    {
        return E_POINTER;
    }

    HENHMETAFILE metafile = CopyEnhMetaFileFromBytes(presentation_.enhancedMetafile);
    if (metafile == nullptr)
    {
        return E_FAIL;
    }

    RECT rect{bounds->left, bounds->top, bounds->right, bounds->bottom};
    BOOL played = PlayEnhMetaFile(drawContext, metafile, &rect);
    DeleteEnhMetaFile(metafile);
    return played ? S_OK : HResultFromWin32LastError();
}

STDMETHODIMP FormulaOleObject::GetColorSet(DWORD, LONG, void*, DVTARGETDEVICE*, HDC, LOGPALETTE** colorSet)
{
    if (colorSet == nullptr)
    {
        return E_POINTER;
    }

    *colorSet = nullptr;
    return S_FALSE;
}

STDMETHODIMP FormulaOleObject::Freeze(DWORD, LONG, void*, DWORD* freeze)
{
    if (freeze == nullptr)
    {
        return E_POINTER;
    }

    *freeze = 0;
    return E_NOTIMPL;
}

STDMETHODIMP FormulaOleObject::Unfreeze(DWORD)
{
    return E_NOTIMPL;
}

STDMETHODIMP FormulaOleObject::SetAdvise(DWORD aspects, DWORD advf, IAdviseSink* adviseSink)
{
    viewAdviseAspects_ = aspects;
    viewAdviseFlags_ = advf;
    viewAdviseSink_ = adviseSink;
    return S_OK;
}

STDMETHODIMP FormulaOleObject::GetAdvise(DWORD* aspects, DWORD* advf, IAdviseSink** adviseSink)
{
    if (aspects != nullptr)
    {
        *aspects = viewAdviseAspects_;
    }

    if (advf != nullptr)
    {
        *advf = viewAdviseFlags_;
    }

    if (adviseSink != nullptr)
    {
        return viewAdviseSink_.CopyTo(adviseSink);
    }

    return S_OK;
}

STDMETHODIMP FormulaOleObject::GetClassID(CLSID* classId)
{
    if (classId == nullptr)
    {
        return E_POINTER;
    }

    *classId = CLSID_LaTeXSnipperFormula;
    return S_OK;
}

STDMETHODIMP FormulaOleObject::IsDirty()
{
    return dirty_ ? S_OK : S_FALSE;
}

STDMETHODIMP FormulaOleObject::InitNew(IStorage* storage)
{
    HRESULT result = SavePresentationToStorage(storage, presentation_);
    if (SUCCEEDED(result))
    {
        dirty_ = false;
    }

    return result;
}

STDMETHODIMP FormulaOleObject::Load(IStorage* storage)
{
    FormulaPresentation loaded;
    HRESULT result = LoadPresentationFromStorage(storage, &loaded);
    if (SUCCEEDED(result))
    {
        presentation_ = std::move(loaded);
        dirty_ = false;
    }

    return SUCCEEDED(result) ? S_OK : result;
}

STDMETHODIMP FormulaOleObject::Save(IStorage* storage, BOOL)
{
    HRESULT result = SavePresentationToStorage(storage, presentation_);
    if (SUCCEEDED(result))
    {
        dirty_ = false;
    }

    return result;
}

STDMETHODIMP FormulaOleObject::SaveCompleted(IStorage*)
{
    return S_OK;
}

STDMETHODIMP FormulaOleObject::HandsOffStorage()
{
    return S_OK;
}

STDMETHODIMP FormulaClassFactory::QueryInterface(REFIID iid, void** object)
{
    if (object == nullptr)
    {
        return E_POINTER;
    }

    if (iid == IID_IUnknown || iid == IID_IClassFactory)
    {
        *object = static_cast<IClassFactory*>(this);
        AddRef();
        return S_OK;
    }

    *object = nullptr;
    return E_NOINTERFACE;
}

STDMETHODIMP_(ULONG) FormulaClassFactory::AddRef()
{
    return static_cast<ULONG>(InterlockedIncrement(&refCount_));
}

STDMETHODIMP_(ULONG) FormulaClassFactory::Release()
{
    const ULONG remaining = static_cast<ULONG>(InterlockedDecrement(&refCount_));
    if (remaining == 0)
    {
        delete this;
    }

    return remaining;
}

STDMETHODIMP FormulaClassFactory::CreateInstance(IUnknown* outer, REFIID iid, void** object)
{
    WriteNativeOleLog(L"ClassFactory CreateInstance entered.");
    if (object == nullptr)
    {
        return E_POINTER;
    }

    *object = nullptr;
    if (outer != nullptr)
    {
        WriteNativeOleLog(L"ClassFactory rejected aggregation.");
        return CLASS_E_NOAGGREGATION;
    }

    FormulaOleObject* formulaObject = new (std::nothrow) FormulaOleObject();
    if (formulaObject == nullptr)
    {
        WriteNativeOleLog(L"FormulaOleObject allocation failed.");
        return E_OUTOFMEMORY;
    }

    HRESULT queryResult = formulaObject->QueryInterface(iid, object);
    WriteNativeOleLog(SUCCEEDED(queryResult) ? L"ClassFactory QueryInterface succeeded." : L"ClassFactory QueryInterface failed.");
    formulaObject->Release();
    return queryResult;
}

STDMETHODIMP FormulaClassFactory::LockServer(BOOL lock)
{
    if (lock)
    {
        InterlockedIncrement(&g_lockCount);
    }
    else
    {
        InterlockedDecrement(&g_lockCount);
    }

    return S_OK;
}

File: office_plugin/hosts/OleFormulaObject/FormulaOleObject.cs
using System;
using System.Runtime.InteropServices;
using LaTeXSnipper.OfficePlugin.Abstractions;

namespace LaTeXSnipper.OfficePlugin.OleFormulaObject;

[ComVisible(true)]
[Guid(OleFormulaObjectIds.ClassIdString)]
[ProgId(OleFormulaObjectIds.ProgId)]
[ClassInterface(ClassInterfaceType.None)]
public sealed class FormulaOleObject : IOleObject, IDataObject, IViewObject, IPersistStorage
{
    private readonly OleFormulaPresentation _presentation;
    private IntPtr _clientSite;
    private uint _viewAdviseAspects;
    private uint _viewAdviseFlags;
    private IntPtr _viewAdviseSink;

    public FormulaOleObject()
    {
        OleServerLog.Write("FormulaOleObject constructing.");
        _presentation = OlePlaceholderPresentationFactory.CreateDefault();
        OleServerLog.Write("FormulaOleObject constructed.");
    }

    public IntPtr QueryInterface(Guid riid)
    {
        if (riid == typeof(IOleObject).GUID)
        {
            return Marshal.GetComInterfaceForObject(this, typeof(IOleObject));
        }

        if (riid == typeof(IDataObject).GUID)
        {
            return Marshal.GetComInterfaceForObject(this, typeof(IDataObject));
        }

        if (riid == typeof(IViewObject).GUID)
        {
            return Marshal.GetComInterfaceForObject(this, typeof(IViewObject));
        }

        if (riid == typeof(IPersistStorage).GUID)
        {
            return Marshal.GetComInterfaceForObject(this, typeof(IPersistStorage));
        }

        if (riid == typeof(object).GUID || riid == new Guid("00000000-0000-0000-C000-000000000046"))
        {
            return Marshal.GetIUnknownForObject(this);
        }

        return IntPtr.Zero;
    }

    public int SetClientSite(IntPtr pClientSite)
    {
        ReleaseClientSite();
        _clientSite = pClientSite;
        if (_clientSite != IntPtr.Zero)
        {
            Marshal.AddRef(_clientSite);
        }

        return ComConstants.SOk;
    }

    public int GetClientSite(out IntPtr ppClientSite)
    {
        ppClientSite = _clientSite;
        if (ppClientSite != IntPtr.Zero)
        {
            Marshal.AddRef(ppClientSite);
        }

        return ComConstants.SOk;
    }

    public int SetHostNames(string szContainerApp, string szContainerObj) => ComConstants.SOk;

    public int Close(uint dwSaveOption)
    {
        if (dwSaveOption == ComConstants.OleCloseSaveIfDirty || dwSaveOption == ComConstants.OleCloseNoSave)
        {
            ReleaseClientSite();
            return ComConstants.SOk;
        }

        return ComConstants.SOk;
    }

    public int SetMoniker(uint dwWhichMoniker, IntPtr pmk) => ComConstants.SOk;

    public int GetMoniker(uint dwAssign, uint dwWhichMoniker, out IntPtr ppmk)
    {
        ppmk = IntPtr.Zero;
        return ComConstants.ENotImpl;
    }

    public int InitFromData(IntPtr pDataObject, bool fCreation, uint dwReserved) => ComConstants.SOk;

    public int GetClipboardData(uint dwReserved, out IntPtr ppDataObject)
    {
        ppDataObject = Marshal.GetComInterfaceForObject(this, typeof(IDataObject));
        return ComConstants.SOk;
    }

    public int DoVerb(int iVerb, IntPtr lpmsg, IntPtr pActiveSite, int lindex, IntPtr hwndParent, IntPtr lprcPosRect)
    {
        return ComConstants.SOk;
    }

    public int EnumVerbs(out IntPtr ppEnumOleVerb)
    {
        ppEnumOleVerb = IntPtr.Zero;
        return ComConstants.ENotImpl;
    }

    public int Update() => ComConstants.SOk;

    public int IsUpToDate() => ComConstants.SOk;

    public int GetUserClassID(out Guid pClsid)
    {
        pClsid = OleFormulaObjectIds.ClassId;
        return ComConstants.SOk;
    }

    public int GetUserType(uint dwFormOfType, out IntPtr pszUserType)
    {
        pszUserType = Marshal.StringToCoTaskMemUni(OleFormulaObjectIds.FriendlyName);
        return ComConstants.SOk;
    }

    public int SetExtent(uint dwDrawAspect, ref SIZEL psizel) => ComConstants.SOk;

    public int GetExtent(uint dwDrawAspect, out SIZEL psizel)
    {
        if (dwDrawAspect != ComConstants.DvAspectContent)
        {
            psizel = default;
            return ComConstants.DvEDvaspect;
        }

        psizel = new SIZEL
        {
            Cx = PointsToHimetric(_presentation.Payload.WidthPoints),
            Cy = PointsToHimetric(_presentation.Payload.HeightPoints)
        };
        return ComConstants.SOk;
    }

    public int Advise(IntPtr pAdvSink, out uint pdwConnection)
    {
        pdwConnection = 0;
        return ComConstants.ENotImpl;
    }

    public int Unadvise(uint dwConnection) => ComConstants.ENotImpl;

    public int EnumAdvise(out IntPtr ppenumAdvise)
    {
        ppenumAdvise = IntPtr.Zero;
        return ComConstants.ENotImpl;
    }

    public int GetMiscStatus(uint dwAspect, out uint pdwStatus)
    {
        pdwStatus = ComConstants.OleMiscRecomposeOnResize
            | ComConstants.OleMiscInsideOut
            | ComConstants.OleMiscActivateWhenVisible;
        return ComConstants.SOk;
    }

    public int SetColorScheme(IntPtr pLogpal) => ComConstants.SOk;

    public int GetData(ref FORMATETC pformatetcIn, out STGMEDIUM pmedium)
    {
        pmedium = default;
        int queryResult = QueryGetData(ref pformatetcIn);
        if (queryResult != ComConstants.SOk)
        {
            return queryResult;
        }

        IntPtr hemf = NativeMethods.SetEnhMetaFileBits((uint)_presentation.EnhancedMetafile.Length, _presentation.EnhancedMetafile);
        if (hemf == IntPtr.Zero)
        {
            return ComConstants.EFail;
        }

        pmedium = new STGMEDIUM
        {
            tymed = ComConstants.TymedEnhmf,
            unionmember = hemf,
            pUnkForRelease = IntPtr.Zero
        };
        return ComConstants.SOk;
    }

    public int GetDataHere(ref FORMATETC pformatetc, ref STGMEDIUM pmedium) => ComConstants.ENotImpl;

    public int QueryGetData(ref FORMATETC pformatetc)
    {
        if (pformatetc.dwAspect != ComConstants.DvAspectContent)
        {
            return ComConstants.DvEDvaspect;
        }

        if ((pformatetc.tymed & ComConstants.TymedEnhmf) == 0)
        {
            return ComConstants.DvETymed;
        }

        return pformatetc.cfFormat == ComConstants.CfEnhmetafile
            ? ComConstants.SOk
            : ComConstants.DvEFormatEtc;
    }

    public int GetCanonicalFormatEtc(ref FORMATETC pformatectIn, out FORMATETC pformatetcOut)
    {
        pformatetcOut = pformatectIn;
        pformatetcOut.ptd = IntPtr.Zero;
        return ComConstants.SFalse;
    }

    public int SetData(ref FORMATETC pformatetc, ref STGMEDIUM pmedium, bool fRelease) => ComConstants.ENotImpl;

    public int EnumFormatEtc(uint dwDirection, out IntPtr ppenumFormatEtc)
    {
        ppenumFormatEtc = IntPtr.Zero;
        return ComConstants.ENotImpl;
    }

    public int DAdvise(ref FORMATETC pformatetc, uint advf, IntPtr pAdvSink, out uint pdwConnection)
    {
        pdwConnection = 0;
        return ComConstants.ENotImpl;
    }

    public int DUnadvise(uint dwConnection) => ComConstants.ENotImpl;

    public int EnumDAdvise(out IntPtr ppenumAdvise)
    {
        ppenumAdvise = IntPtr.Zero;
        return ComConstants.ENotImpl;
    }

    public int Draw(uint dwDrawAspect, int lindex, IntPtr pvAspect, IntPtr ptd, IntPtr hdcTargetDev, IntPtr hdcDraw, IntPtr lprcBounds, IntPtr lprcWBounds, IntPtr pfnContinue, IntPtr dwContinue)
    {
        if (dwDrawAspect != ComConstants.DvAspectContent)
        {
            return ComConstants.DvEDvaspect;
        }

        if (hdcDraw == IntPtr.Zero || lprcBounds == IntPtr.Zero)
        {
            return ComConstants.EPointer;
        }

        IntPtr hemf = NativeMethods.SetEnhMetaFileBits((uint)_presentation.EnhancedMetafile.Length, _presentation.EnhancedMetafile);
        if (hemf == IntPtr.Zero)
        {
            return ComConstants.EFail;
        }

        try
        {
            RECT bounds = Marshal.PtrToStructure<RECT>(lprcBounds);
            return NativeMethods.PlayEnhMetaFile(hdcDraw, hemf, ref bounds) ? ComConstants.SOk : ComConstants.EFail;
        }
        finally
        {
            NativeMethods.DeleteEnhMetaFile(hemf);
        }
    }

    public int GetColorSet(uint dwDrawAspect, int lindex, IntPtr pvAspect, IntPtr ptd, IntPtr hicTargetDev, out IntPtr ppColorSet)
    {
        ppColorSet = IntPtr.Zero;
        return ComConstants.SFalse;
    }

    public int Freeze(uint dwDrawAspect, int lindex, IntPtr pvAspect, out uint pdwFreeze)
    {
        pdwFreeze = 0;
        return ComConstants.ENotImpl;
    }

    public int Unfreeze(uint dwFreeze) => ComConstants.ENotImpl;

    public int SetAdvise(uint aspects, uint advf, IntPtr pAdvSink)
    {
        _viewAdviseAspects = aspects;
        _viewAdviseFlags = advf;
        _viewAdviseSink = pAdvSink;
        return ComConstants.SOk;
    }

    public int GetAdvise(out uint pAspects, out uint pAdvf, out IntPtr ppAdvSink)
    {
        pAspects = _viewAdviseAspects;
        pAdvf = _viewAdviseFlags;
        ppAdvSink = _viewAdviseSink;
        if (ppAdvSink != IntPtr.Zero)
        {
            Marshal.AddRef(ppAdvSink);
        }

        return ComConstants.SOk;
    }

    public int GetClassID(out Guid pClassID)
    {
        pClassID = OleFormulaObjectIds.ClassId;
        return ComConstants.SOk;
    }

    public int IsDirty() => ComConstants.SFalse;

    public int InitNew(IntPtr pStg) => ComConstants.SOk;

    public int Load(IntPtr pStg) => ComConstants.SOk;

    public int Save(IntPtr pStgSave, bool fSameAsLoad) => ComConstants.SOk;

    public int SaveCompleted(IntPtr pStgNew) => ComConstants.SOk;

    public int HandsOffStorage() => ComConstants.SOk;

    private void ReleaseClientSite()
    {
        if (_clientSite != IntPtr.Zero)
        {
            Marshal.Release(_clientSite);
            _clientSite = IntPtr.Zero;
        }
    }

    private static int PointsToHimetric(double points)
    {
        return (int)Math.Ceiling(points * 2540d / 72d);
    }
}
